File: gfs-client/gfs-client.py
'''
Definition of BaseGFSClient
class BaseGFSClient:
    def readChunk(self, filename, chunkIndex):
        # Read a chunk from GFS
    def writeChunk(self, filename, chunkIndex, content):
        # Write a chunk to GFS
'''
class GFSClient(BaseGFSClient):

    # @param {int} chunkSize chunk size bytes
    def __init__(self, chunkSize):
        BaseGFSClient.__init__(self)
        # initialize your data structure here
        self.file_chunk = {}
        self.chunk_size = chunkSize

    # @param {str} filename a file name
    # @return {str} conetent of the file given from GFS
    def read(self, filename):
        # Write your code here
        if filename in self.file_chunk:
            n = self.file_chunk[filename]
        else:
            return None
        ans = ""
        for i in xrange(n):
            ans += BaseGFSClient.readChunk(self, filename, i)
        return ans

    # @param {str} filename a file name
    # @param {str} content a string
    # @return nothing
    def write(self, filename, content):
        # Write your code here
        le = len(content)
        for i in xrange(le / self.chunk_size):
            BaseGFSClient.writeChunk(self, filename, i, content[i * self.chunk_size: (i + 1) * self.chunk_size])
        self.file_chunk[filename] = le / self.chunk_size
        if le % self.chunk_size:
            BaseGFSClient.writeChunk(self, filename, le / self.chunk_size, content[le / self.chunk_size * self.chunk_size:])
            self.file_chunk[filename] += 1

File: gfs-client/gfs-client.cc
/* Definition of BaseGFSClient
 * class BaseGFSClient {
 * private: map<string, string> chunk_list;
 * public:
 *     string readChunk(string& filename, int chunkIndex) {
 *         // Read a chunk from GFS
 *     }
 *     void writeChunk(string& filename, int chunkIndex,
 *                     string& content) {
 *         // Write a chunk to GFS     
 *     }
 *  };
 */
class GFSClient : public BaseGFSClient {
public:
    GFSClient(int chunkSize) {
        // initialize your data structure here
        chunk_size = chunkSize;
        file_chunks.clear();
    }

    // @param filename a file name
    // @return conetent of the file given from GFS
    string read(string& filename) {
        // Write your code here
        string ans;
        if (file_chunks.find(filename) != file_chunks.end()) {
            int cnt = file_chunks[filename];
            for (int i = 0; i < cnt; ++i) {
                ans += readChunk(filename, i);
            }
        }
        return ans;
    }

    // @param filename a file name
    // @param content a string
    // @return void
    void write(string& filename, string& content) {
        // Write your code here
        int len = content.length(), i = 0;
        while (i < len / chunk_size) {
            string t = content.substr(i * chunk_size, chunk_size);
            writeChunk(filename, i, t);
            ++i;
        }
        if (len % chunk_size) {
            string t = content.substr(i * chunk_size, len % chunk_size);
            writeChunk(filename, i, t);
            ++i;
        }
        file_chunks[filename] = i;
    }

private:
    int chunk_size;
    map<string, int> file_chunks;
};

